enclave-runtime/src/initialization.rs

```diff
@@ -0,0 +1,264 @@
+	env_logger::init();
+
+	ed25519::create_sealed_if_absent().map_err(Error::Crypto)?;
+	let signer = Ed25519Seal::unseal().map_err(Error::Crypto)?;
+	info!("[Enclave initialized] Ed25519 prim raw : {:?}", signer.public().0);
+
+	rsa3072::create_sealed_if_absent()?;
+
+	// Create the aes key that is used for state encryption such that a key is always present in tests.
+	// It will be overwritten anyway if mutual remote attastation is performed with the primary worker.
+	aes::create_sealed_if_absent().map_err(Error::Crypto)?;
+
+	let state_handler = Arc::new(GlobalFileStateHandler);
+	GLOBAL_STATE_HANDLER_COMPONENT.initialize(state_handler.clone());
+
+	let ocall_api = Arc::new(OcallApi);
+	GLOBAL_OCALL_API_COMPONENT.initialize(ocall_api.clone());
+
+	let stf_executor = Arc::new(EnclaveStfExecutor::new(ocall_api.clone(), state_handler.clone()));
+	GLOBAL_STF_EXECUTOR_COMPONENT.initialize(stf_executor);
+
+	// For debug purposes, list shards. no problem to panic if fails.
+	let shards = state_handler.list_shards().unwrap();
+	debug!("found the following {} shards on disk:", shards.len());
+	for s in shards {
+		debug!("{}", s.encode().to_base58())
+	}
+
+	itp_primitives_cache::set_primitives(
+		GLOBAL_PRIMITIVES_CACHE.as_ref(),
+		mu_ra_url,
+		untrusted_worker_url,
+	)
+	.map_err(Error::PrimitivesAccess)?;
+
+	let shielding_key = Rsa3072Seal::unseal()?;
+	let watch_extractor = Arc::new(create_determine_watch::<Hash>());
+	let connection_registry = Arc::new(ConnectionRegistry::<Hash, TungsteniteWsConnection>::new());
+
+	// We initialize components for the public RPC / direct invocation server here, so we can start the server
+	// before registering on the parentchain. If we started the RPC AFTER registering on the parentchain and
+	// initializing the light-client, there is a period of time where a peer might want to reach us,
+	// but the RPC server is not yet up and running, resulting in error messages or even in that
+	// validateer completely breaking (IO PipeError).
+	// Corresponding GH issues are #545 and #600.
+
+	let rpc_author = its_sidechain::top_pool_rpc_author::initializer::create_top_pool_rpc_author(
+		connection_registry.clone(),
+		state_handler,
+		ocall_api,
+		shielding_key,
+	);
+	GLOBAL_RPC_AUTHOR_COMPONENT.initialize(rpc_author.clone());
+
+	let io_handler = public_api_rpc_handler(rpc_author);
+	let rpc_handler = Arc::new(RpcWsHandler::new(io_handler, watch_extractor, connection_registry));
+	GLOBAL_RPC_WS_HANDLER_COMPONENT.initialize(rpc_handler);
+
+	let sidechain_block_import_queue = Arc::new(EnclaveSidechainBlockImportQueue::default());
+	GLOBAL_SIDECHAIN_IMPORT_QUEUE_COMPONENT.initialize(sidechain_block_import_queue);
+
+	Ok(())
+}
+
+pub(crate) fn init_enclave_sidechain_components() -> EnclaveResult<()> {
+	let stf_executor = GLOBAL_STF_EXECUTOR_COMPONENT.get().ok_or_else(|| {
+		error!("Failed to retrieve global STF executor component (maybe it is not initialized?)");
+		Error::ComponentNotInitialized
+	})?;
```

Comment:
Yes, agreed, this can be simplified and the amount of logging reduced 👍  thanks for pointing that out. Now I wish I had a `nameof()` function in Rust too, so I wouldn't need to type all the variable names as string for the errors 🙈

Reply:
I agree to that. We could introduce an enum for the components and implement `toString` for that, but there might be a better way.

Reply:
I think refactoring the `get` method of `ComponentContainer` might be sensible, to return an error that already contains all the information. And a container can also have a name that it puts into the error itself, so the caller does not have to care about the error contents.

Reply:
The component container now knows its own name, and the getter returns a proper `Result` with an error that includes its name. That allowed me to reduce a lot of the getter calls to just `.get()?;`, which is certainly a lot cleaner. Thanks for insisting 😉

Reply:
Coolio! Thanks for the extra effort. I think this was worth it!